File: stripe-captured-no-order-created/python/reconcile_orphaned_captures.py
"""Find Stripe PaymentIntents that captured money with no matching Medusa order,
and repair them the safe way. Never inserts a synthetic order through the Admin
API. DRY_RUN=true only logs the reconciliation records it would act on. The one
safe repair is retrying POST /store/carts/{id}/complete, the same route the
storefront already calls, since completeCartWorkflow's idempotent flag was set
to false in Medusa v2.8.0 specifically so a stalled completion can be retried.

Guide: https://www.allanninal.dev/medusa/stripe-captured-no-order-created/
"""
import os
import time
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_orphaned_captures")

STRIPE_KEY = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
STRIPE_API = "https://api.stripe.com/v1"

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
PUBLISHABLE_KEY = os.environ.get("MEDUSA_PUBLISHABLE_KEY", "pk_dummy")

GRACE_MS = float(os.environ.get("GRACE_MINUTES", "10")) * 60 * 1000
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def decide_reconciliation(
    stripe_payment_intent_id,
    stripe_status,
    captured_at_ms,
    now_ms,
    grace_ms,
    medusa_payment_data_ids,
    cart_completed_at,
    cart_has_order_id,
):
    """Pure: no I/O. Returns one of ok, too_recent,
    orphaned_capture_needs_manual_complete, already_reconciled.
    """
    matched_in_medusa = stripe_payment_intent_id in medusa_payment_data_ids

    if matched_in_medusa and (cart_completed_at is not None or cart_has_order_id):
        return "already_reconciled"

    if stripe_status != "succeeded":
        return "ok"  # nothing captured yet, not our problem

    if now_ms - captured_at_ms < grace_ms:
        return "too_recent"  # webhook may still be in flight, don't flag yet

    if not matched_in_medusa and cart_completed_at is None and not cart_has_order_id:
        return "orphaned_capture_needs_manual_complete"

    return "ok"


def recent_succeeded_payment_intents(lookback_hours=24):
    out, starting_after = [], None
    since = int(time.time()) - lookback_hours * 3600
    while True:
        params = {"limit": 100, "created[gte]": since}
        if starting_after:
            params["starting_after"] = starting_after
        r = requests.get(
            f"{STRIPE_API}/payment_intents",
            params=params,
            auth=(STRIPE_KEY, ""),
            timeout=30,
        )
        r.raise_for_status()
        body = r.json()
        for pi in body["data"]:
            if pi["status"] == "succeeded":
                out.append(pi)
        if not body.get("has_more"):
            return out
        starting_after = body["data"][-1]["id"]


def get_admin_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def all_medusa_payment_data_ids(token):
    headers = {"Authorization": f"Bearer {token}"}
    ids, offset, limit = [], 0, 100
    while True:
        r = requests.get(
            f"{BASE_URL}/admin/payments",
            params={"fields": "id,data", "limit": limit, "offset": offset},
            headers=headers,
            timeout=30,
        )
        r.raise_for_status()
        body = r.json()
        for payment in body["payments"]:
            pid = (payment.get("data") or {}).get("id")
            if pid:
                ids.append(pid)
        offset += limit
        if offset >= body["count"]:
            return ids


def get_cart(cart_id):
    r = requests.get(
        f"{BASE_URL}/store/carts/{cart_id}",
        headers={"x-publishable-api-key": PUBLISHABLE_KEY},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["cart"]


def complete_cart(cart_id):
    r = requests.post(
        f"{BASE_URL}/store/carts/{cart_id}/complete",
        headers={"x-publishable-api-key": PUBLISHABLE_KEY},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("type") == "order":
        return body["order"]
    raise RuntimeError(f"cart {cart_id} did not complete into an order: {body}")


def run():
    token = get_admin_token()
    medusa_payment_data_ids = all_medusa_payment_data_ids(token)
    payment_intents = recent_succeeded_payment_intents()
    now_ms = time.time() * 1000

    flagged = []
    for pi in payment_intents:
        cart_id = (pi.get("metadata") or {}).get("cart_id")
        if not cart_id:
            continue
        cart = get_cart(cart_id)
        outcome = decide_reconciliation(
            stripe_payment_intent_id=pi["id"],
            stripe_status=pi["status"],
            captured_at_ms=pi["created"] * 1000,
            now_ms=now_ms,
            grace_ms=GRACE_MS,
            medusa_payment_data_ids=medusa_payment_data_ids,
            cart_completed_at=cart.get("completed_at"),
            cart_has_order_id=bool(cart.get("order")),
        )
        if outcome == "orphaned_capture_needs_manual_complete":
            flagged.append((pi, cart_id))

    if not flagged:
        log.info("No orphaned captures found across %d succeeded PaymentIntent(s).", len(payment_intents))
        return

    for pi, cart_id in flagged:
        log.warning(
            "Orphaned capture: PI %s amount=%s cart=%s captured_at=%s. %s",
            pi["id"], pi["amount"], cart_id, pi["created"],
            "Would retry cart complete" if DRY_RUN else "Retrying cart complete",
        )
        if DRY_RUN:
            continue

        fresh_cart = get_cart(cart_id)
        if fresh_cart.get("completed_at") or fresh_cart.get("order"):
            log.info("Cart %s completed between detection and repair. Skipping.", cart_id)
            continue

        try:
            order = complete_cart(cart_id)
            log.info("Cart %s completed into order %s.", cart_id, order.get("id"))
        except Exception as exc:
            log.error(
                "Cart %s failed to complete for PI %s: %s. "
                "Flagging to support for manual /admin/draft-orders reconciliation.",
                cart_id, pi["id"], exc,
            )

    log.info("Done. %d orphaned capture(s) %s.", len(flagged), "to review" if DRY_RUN else "processed")


if __name__ == "__main__":
    run()

File: stripe-captured-no-order-created/node/reconcile-orphaned-captures.js
/**
 * Find Stripe PaymentIntents that captured money with no matching Medusa order,
 * and repair them the safe way. Never inserts a synthetic order through the
 * Admin API. DRY_RUN=true only logs the reconciliation records it would act
 * on. The one safe repair is retrying POST /store/carts/{id}/complete, the
 * same route the storefront already calls, since completeCartWorkflow's
 * idempotent flag was set to false in Medusa v2.8.0 specifically so a stalled
 * completion can be retried.
 *
 * Guide: https://www.allanninal.dev/medusa/stripe-captured-no-order-created/
 */
import { pathToFileURL } from "node:url";

const STRIPE_KEY = process.env.STRIPE_SECRET_KEY || "sk_test_dummy";
const STRIPE_API = "https://api.stripe.com/v1";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const PUBLISHABLE_KEY = process.env.MEDUSA_PUBLISHABLE_KEY || "pk_dummy";

const GRACE_MS = Number(process.env.GRACE_MINUTES || 10) * 60 * 1000;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function decideReconciliation({
  stripePaymentIntentId,
  stripeStatus,
  capturedAtMs,
  nowMs,
  graceMs,
  medusaPaymentDataIds,
  cartCompletedAt,
  cartHasOrderId,
}) {
  const matchedInMedusa = medusaPaymentDataIds.includes(stripePaymentIntentId);

  if (matchedInMedusa && (cartCompletedAt !== null || cartHasOrderId)) {
    return "already_reconciled";
  }
  if (stripeStatus !== "succeeded") {
    return "ok"; // nothing captured yet, not our problem
  }
  if (nowMs - capturedAtMs < graceMs) {
    return "too_recent"; // webhook may still be in flight, don't flag yet
  }
  if (!matchedInMedusa && cartCompletedAt === null && !cartHasOrderId) {
    return "orphaned_capture_needs_manual_complete";
  }
  return "ok";
}

async function recentSucceededPaymentIntents(lookbackHours = 24) {
  const out = [];
  let startingAfter = null;
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  while (true) {
    const params = new URLSearchParams({ limit: "100", "created[gte]": String(since) });
    if (startingAfter) params.set("starting_after", startingAfter);
    const res = await fetch(`${STRIPE_API}/payment_intents?${params}`, {
      headers: { Authorization: `Basic ${Buffer.from(STRIPE_KEY + ":").toString("base64")}` },
    });
    if (!res.ok) throw new Error(`Stripe ${res.status}`);
    const body = await res.json();
    for (const pi of body.data) if (pi.status === "succeeded") out.push(pi);
    if (!body.has_more) return out;
    startingAfter = body.data[body.data.length - 1].id;
  }
}

async function getAdminToken() {
  const res = await fetch(`${BASE_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: EMAIL, password: PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  return (await res.json()).token;
}

async function allMedusaPaymentDataIds(token) {
  const ids = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const res = await fetch(
      `${BASE_URL}/admin/payments?fields=id,data&limit=${limit}&offset=${offset}`,
      { headers: { Authorization: `Bearer ${token}` } }
    );
    if (!res.ok) throw new Error(`Medusa ${res.status}`);
    const body = await res.json();
    for (const payment of body.payments) {
      const pid = payment.data?.id;
      if (pid) ids.push(pid);
    }
    offset += limit;
    if (offset >= body.count) return ids;
  }
}

async function getCart(cartId) {
  const res = await fetch(`${BASE_URL}/store/carts/${cartId}`, {
    headers: { "x-publishable-api-key": PUBLISHABLE_KEY },
  });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  return (await res.json()).cart;
}

async function completeCart(cartId) {
  const res = await fetch(`${BASE_URL}/store/carts/${cartId}/complete`, {
    method: "POST",
    headers: { "x-publishable-api-key": PUBLISHABLE_KEY },
  });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  const body = await res.json();
  if (body.type === "order") return body.order;
  throw new Error(`cart ${cartId} did not complete into an order: ${JSON.stringify(body)}`);
}

export async function run() {
  const token = await getAdminToken();
  const medusaPaymentDataIds = await allMedusaPaymentDataIds(token);
  const paymentIntents = await recentSucceededPaymentIntents();
  const nowMs = Date.now();

  const flagged = [];
  for (const pi of paymentIntents) {
    const cartId = pi.metadata?.cart_id;
    if (!cartId) continue;
    const cart = await getCart(cartId);
    const outcome = decideReconciliation({
      stripePaymentIntentId: pi.id,
      stripeStatus: pi.status,
      capturedAtMs: pi.created * 1000,
      nowMs,
      graceMs: GRACE_MS,
      medusaPaymentDataIds,
      cartCompletedAt: cart.completed_at ?? null,
      cartHasOrderId: Boolean(cart.order),
    });
    if (outcome === "orphaned_capture_needs_manual_complete") {
      flagged.push([pi, cartId]);
    }
  }

  if (flagged.length === 0) {
    console.log(`No orphaned captures found across ${paymentIntents.length} succeeded PaymentIntent(s).`);
    return;
  }

  for (const [pi, cartId] of flagged) {
    console.warn(
      `Orphaned capture: PI ${pi.id} amount=${pi.amount} cart=${cartId} captured_at=${pi.created}. ${DRY_RUN ? "Would retry cart complete" : "Retrying cart complete"}`
    );
    if (DRY_RUN) continue;

    const freshCart = await getCart(cartId);
    if (freshCart.completed_at || freshCart.order) {
      console.log(`Cart ${cartId} completed between detection and repair. Skipping.`);
      continue;
    }

    try {
      const order = await completeCart(cartId);
      console.log(`Cart ${cartId} completed into order ${order.id}.`);
    } catch (err) {
      console.error(
        `Cart ${cartId} failed to complete for PI ${pi.id}: ${err.message}. Flagging to support for manual /admin/draft-orders reconciliation.`
      );
    }
  }

  console.log(`Done. ${flagged.length} orphaned capture(s) ${DRY_RUN ? "to review" : "processed"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
